## Implementing Multiple Linear Regression Using Gradient Descent And Stochastic Gradient Descent Algorithm.

In this assignment, you will be implementing Multiple Linear Regression Model on the same data set as in programming_assignment_1b, but here we will be using gradient descent algorithm (GDA) and stochastic gradient descent algorithm (SGDA) to minimize the cost function as taught in the class. The SGDA implementation is optional.

Please add your own print statements to check your code to ensure your code is correct in every step.  (Note: we will not be grading the print statements you add to your code.)

In [385]:
# run this code cell using shift+enter before moving further
import numpy as np
import pandas as pd
import matplotlib
from matplotlib import pyplot as plt

# Loading the data set.

In the below code cell, you will load the data using python pandas library as done in the programming_assignment_1b. Use pd.read_csv('File url ', header=None,.... ) with the value of header=None,delim_whitespace=True,names=names,na_values='?' as attributes. The url for the .data file is https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data. This step is same as done in programming_assignment_1b.

In [386]:
# After completing the code in this code cell, run this code cell before moving further.
names =[
    'CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 
    'AGE',  'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'PRICE'
]

#Write your code below to save dataframe in the df variable below. 
# In place of None, write the pandas command to read the csv file.
df= pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/housing/housing.data', header = None, delim_whitespace = True, names = names, na_values = '?')


Now, instead of taking just 1 feature i.e 'RM' as we did in Simple Linear Regression, we will use 4 features which are 'LSTAT', 'DIS', 'RM' and 'INDUS'. We choose only these features because they are almost on similar scale. So, there is no need for feature scaling. Fetching the values of 5 columns into a smaller dataframe (say) df1 from df. This step is also same as done in programming_assignment_1b.There we fetched only 2 columns but here we need 5 columns.We need the values in the 'PRICE', 'LSTAT', 'DIS', 'RM' and 'INDUS' columns.

In [408]:
#  After completing the code in this code cell, run this code cell before moving further. Also, save the values in df2 after dropping empty rows from df1
# Write your code below.
df1=df[['LSTAT', 'DIS', 'RM', 'INDUS', 'PRICE']]
df2=df1.dropna()
print(df2.shape)
# Check the shape of df2. It should be (506,5)


(506, 5)


Create a vector y having the values of 'PRICE' column and matrix x having the values of  'LSTAT', 'DIS', 'RM' and 'INDUS' columns. We have already written the code for that. Just run the code cell below.

In [410]:
# Just run this code cell
df3 = df2[['LSTAT', 'DIS', 'RM', 'INDUS']]
x=df3.values
y=np.array(df2['PRICE'])
# Check the shape of x and y vectors.
print(x.shape)
print(y.shape)

(506, 4)
(506,)


Reshape y to be rank 2. After checking the shape of x and y in the above code cell, we see that x is already rank 2 and y are rank 1 matrix. Before moving ahead, convert y to be rank 2 matrix. For example, I would use the command y=y.reshape(y.shape[0],1) to reshape.

In [411]:
# After completing the code in this code cell, run this code cell before moving further. 
# Write your code below
y = y.reshape(y.shape[0],1)


Calculating the value of n i.e. number of training examples. 
Hint: Value of n is equal to the number of rows in either x or y matrix which can be accessed using numpy shape command.

In [412]:
# After completing the code in this code cell, run this code cell before moving further. 
# Write your code below
n = y.shape[0]
# After writing a code, it is a good practice to verify that your code is correct. 
# For example, in this case you can print the value of n and check that it should be equal to 506.
# Occasionally we created test code cell to verify your code. 
# However, you should do it for almost every code you write.

# Appending a Column of ones in x: Just run the code cell below
Warning: Just run the below code cell only once or you would be appending column of ones multiple times if you run the cell multiple times. If by mistake you run the code cell more than once, re-initilize the value of x again in the previous code cell.

In [413]:
a=np.ones((y.shape[0],1))
x=np.hstack((a , x))

In [414]:
# Shape of x should be (506,5)
print(x.shape)

(506, 5)


# Cost Function.
Compute the cost: Write the code to compute the cost inside the function. Do not change the function name or function parameters.

In [415]:
def compute_cost(x, y, beta, n):
    #Write your code in place of None. Cost can be calculated using a single line of code.
    # Remember beta is a vector here.
    cost = np.sum((np.dot(x,beta) - y)**2) / (2*n)
    
    return cost

Before moving ahead, ensure that the code you have written to compute the cost is correct. Just run the below cell.

In [421]:
beta_testcase=np.zeros((5,1))
cost_verify= compute_cost(x, y, beta_testcase, n)

print(cost_verify)

296.073458498


Your output should be equal to 296.073458498. If it's equal to this, then move ahead. Else, re-check your code and re-verify.

# Gradient Descent
Write the code to perform gradient descent in the function below

In [398]:
def gradient_descent(x , y , learning_rate , beta , n , num_iters):
    # In place of None, write the updated value of beta0 in temp0 and of beta1 in temp1
    X = x.T
    for i in range(num_iters):
        
        # derivative vector is given by : X_train.Transpose *  (( X_train * betavector)- y)
        AB_y = x.dot(beta) - y
        #temp =  beta[i] - (learning_rate/n) * np.sum((beta[0] + x[i].dot(beta[i]) - y[i]) * x.T[i])
        beta =  beta - (learning_rate/n) * X.dot(AB_y)
        
        
        if(i%100==0):
            # In place of None, call the cost you just coded above
            cost= compute_cost(x,y,beta,n)
            #print("Cost")
            #print(cost)
             
            
            
    return beta      

Before moving ahead, ensure that your code to update beta is correct. Run the code cell below.

In [419]:
beta_testcase=np.zeros((5,1))
g=gradient_descent(x , y , 0.0049 , beta_testcase , n , 100000)
print(g[0])


[ 7.46236374]


# Integrating the Batch Gradient Descent Function 

Integrating the above function into a single function multiple_linear_reg_model_gda: This function uses gradient descent algorithm to minimize the cost.

In [429]:
def multiple_linear_reg_model_gda(x , y , n , learning_rate , num_iters):
    #initialize the values of parameter vector beta. It should be a column vector of zeros of dimension(n,1)
    beta = np.zeros((5,1))
    
    #calculate the initial cost by calling the function you coded above.
    initial_cost= compute_cost(x,y,beta,n)
    print("Initial Cost")
    print(initial_cost)
    
    #calculate the optimized value of gradients by calling the gradient_descent function coded above
    
    beta = gradient_descent(x , y , learning_rate , beta , n , num_iters)
    
    #Calculate the cost with the optimized value of beta0 and beta1 by calling the cost function.
    
    final_cost = compute_cost(x,y,beta,n)
    print("Final Cost")
    print(final_cost)
    return beta

Now, when you have coded multiple_linear_reg_model_gda function, you can call this function to find the optimized values of parameters beta. Before calling the function, set the values of learning_rate and num_iters. You may have to call this function several number of times with different values of num_iters and learning_rate to find the optimal values of beta. For some values of learning_rate, it may give an error as the values of cost may reach a very high value(infinity) due to overshooting as discussed in the class. 

In [441]:
# Write your code below
learning_rate = 0.005
num_iters = 1000
# In place of None, call the multiple_linear_reg_model_gda.
beta = multiple_linear_reg_model_gda(x , y , n , learning_rate , num_iters)

# The value of final cost should be 14.3470049896 or nearly this(depending on the values of learning_rate and num_itersations you choose.)

Initial Cost
296.073458498
Final Cost
475817512.044


# Normal Equation Method
Now, we will be writing the code to find the values of parameters beta for our multiple linear regression model. This can also be used to cross-check the optimal values of beta we just found above using method above. These values should be same (or nearly same).
Instead of writing the code for normal equation in one line, you can break this into 3 parts: First calculate q=inverse of (dot of (X.T,X)) (these are pseudo commands, use original numpy commands to calculate q). Then w= dot of ( X.T , y) and then beta_vec= dot of (q,w). Here, beta_vec is vector of dimension (5,1) having two values. Example beta0=beta_vec[0][0].

Note : Do not append a column of ones in x because you have already done before. 

In [442]:
# Write the code below
q = np.linalg.inv(np.dot(x.T,x))
w = np.dot(x.T,y)
beta_vec = np.dot(q,w)


In [443]:
print(beta_vec)

[[ 8.29647186]
 [-0.66467285]
 [-0.90387731]
 [ 4.58064345]
 [-0.24342177]]


The Values of beta you just got above should be approximately same as the ones you got using multiple_linear_reg_model_gda.
This assignment ends here.

# Stochastic Gradient Descent (Optional / Ungraded)

In [ ]:
def stochastic_gradient_descent(x , y , learning_rate , beta , n , num_iters):

    for j in range(num_iters):
    
        for i in range(0,n):
        # Write updated value of beta vector in place of None
            temp = None 
        
            beta = temp
   

        if(j%2000==0):
            cost= None
            print("Cost")
            print(cost)
            
            
    return beta

In [ ]:
def multiple_linear_reg_model_sgda(x , y , n , learning_rate , num_iters):
    #initialize the values of parameter vector beta. It should be a column vector of zeros of dimension(m,1)
    beta= None
    
    #calculate the initial cost by calling the function you coded above.
    initial_cost=None
    print("Initial Cost")
    print(initial_cost)
    
    #calculate the optimized value of gradients by calling the stochastic_gradient_descent function coded above
    
    beta= None
    
    #Calculate the cost with the optimized value of beta by calling the cost function.
    
    final_cost=None
    print("Final Cost")
    print(final_cost)
    return beta

Now, when you have coded multiple_linear_reg_model_sgda function, you can call this function to find the optimized values of parameters beta. Before calling the function, set the values of learning_rate and num_iters. You may have to call this function several number of times with different values of num_iters and learning_rate to find the optimal values of parameters beta. For some values of learning_rate, it may give an error as the values of parameters may reach a very high value(infinity).

In [ ]:
# Write your code below
learning_rate=None
num_iters=None
# In place of None, call the function multiple_linear_reg_model_sgda.
beta=None